packages/gatsby/src/query/query-compiler.js

```diff
@@ -1,384 +1,400 @@
+      const otherQuery = processedQueries.get(filePath)
+
+      addError(
+        multipleRootQueriesError(
+          filePath,
+          originalDefinition.def,
+          otherQuery && definitionsByName.get(otherQuery.name).def
         )
       )
-    } catch (error) {
-      const { formattedMessage, docName, message, codeBlock } = graphqlError(
-        namePathMap,
-        nameDefMap,
-        error
-      )
-      nameErrorMap.set(docName, { formattedMessage, message, codeBlock })
-      boundActionCreators.queryExtractionGraphQLError({
-        componentPath: namePathMap.get(docName),
-        error: formattedMessage,
-      })
 
-      const filePath = namePathMap.get(docName)
-      addError(errorParser({ message, filePath }))
+      actions.queryExtractionGraphQLError({
+        componentPath: filePath,
+      })
+      continue
+    }
 
-      return false
+    const usedFragments = new Set()
+    const stack = [operation]
+    let missingFragment = false
+    while (stack.length > 0) {
+      const def = stack.pop(operation)
+      visit(def, {
+        [Kind.FRAGMENT_SPREAD]: node => {
+          const name = node.name.value
+          if (usedFragmentsForFragment.has(name)) {
+            usedFragmentsForFragment.get(name).forEach(derivedFragmentName => {
+              usedFragments.add(derivedFragmentName)
+            })
+            usedFragments.add(name)
+          } else if (definitionsByName.has(name)) {
+            stack.push(definitionsByName.get(name).def)
+            usedFragments.add(name)
+          } else {
+            missingFragment = true
+            actions.queryExtractionGraphQLError({
+              componentPath: filePath,
+            })
+            addError(
+              unknownFragmentError({
+                fragmentNames,
+                filePath,
+                def,
+                node,
+              })
+            )
+          }
+        },
+      })
     }
 
-    // relay-compiler v1.5.0 added "StripUnusedVariablesTransform" to
-    // printTransforms. Unfortunately it currently doesn't detect variables
-    // in input objects widely used in gatsby, and therefore removing
-    // variable declaration from queries.
-    // As a temporary workaround remove that transform by slicing printTransforms.
-    const printContext = printTransforms
-      .slice(0, -1)
-      .reduce((ctx, transform) => transform(ctx, this.schema), compilerContext)
-
-    const fragments = []
-    compilerContext.documents().forEach(node => {
-      if (node.kind === `Fragment`) {
-        fragments.push(node.name)
-      }
-    })
+    if (missingFragment) {
+      continue
+    }
 
-    compilerContext.documents().forEach((node: { name: string }) => {
-      if (node.kind !== `Root`) return
-      const { name } = node
-      let filePath = namePathMap.get(name) || ``
-      if (compiledNodes.has(filePath)) {
-        let otherNode = compiledNodes.get(filePath)
+    const document = {
```

Comment:
Just making sure I understand this.
We prepend all fragments that are used by a query into the graphql document automatically?

Reply:
Yes.